stix2validator/musts.py

```diff
@@ -327,18 +328,71 @@ def types_strict(instance):
                                 % (key, obj['type']), instance['id'])
 
 
-def patterns(instance):
-    """Ensure that the syntax of the pattern of an indicator is valid.
+def patterns(instance, options):
+    """Ensure that the syntax of the pattern of an indicator is valid, and that
+    objects and properties referenced by the pattern are valid.
     """
     if instance['type'] != 'indicator' or 'pattern' not in instance:
         return
 
     pattern = instance['pattern']
     errors = pattern_validator(pattern)
 
-    for e in errors:
-        yield JSONError("Pattern failed to validate: %s."
-                        % e, instance['id'])
+    # Check pattern syntax
+    if errors:
+        for e in errors:
+            yield PatternError(str(e), instance['id'])
+        return
+
+    p = Pattern(pattern)
+    inspection = p.inspect().comparisons
+    for objtype in inspection:
+        # Check observable object types
+        if objtype in enums.OBSERVABLE_TYPES:
+            pass
+        elif options.strict_types:
+            yield PatternError("'%s' is not a valid STIX observable type"
+                               % objtype, instance['id'])
+        elif (not re.match('^\\-?[a-z0-9]+(-[a-z0-9]+)*\\-?$', objtype) or
```

Comment:
Consider pulling out these (and other) regular expressions into pre-compiled RE objects. Giving them a name makes them more descriptive, as well as making them easier to test independently (don't necessarily worry about adding those tests now).